File: Backend/src/mapping/graphBuilder.js
/**
 * Navigation Graph Builder
 * 
 * Converts scan results (rooms + paths) into a navigation graph for pathfinding.
 * 
 * The graph is the single source of truth for navigation.
 * 
 * Requirements:
 * - Nodes: intersections, corridor points, room connections
 * - Edges: walkable connections between nodes
 * - Room-to-node mappings: which node represents each room
 * - A* ready: usable by pathfinding algorithms
 * - Deterministic: same input → same output
 */

const GRAPH_BUILDER_VERSION = '1.0.0';
const COORDINATE_PRECISION = 2;
const NODE_MERGE_THRESHOLD = 5; // Merge nodes within 5 pixels
const ROOM_CORRIDOR_DISTANCE = 50; // Max distance for room-corridor connection
const MIN_EDGE_DISTANCE = 1; // Minimum edge distance

/**
 * Main graph builder entry point
 * @param {Object} scanResult - Output from scanFloorPlan
 * @param {Object} options - Builder options
 * @returns {Object} NavigationGraph
 */
function buildNavigationGraph(scanResult, options = {}) {
  const {
    mergeNearbyNodes = true,
    connectRoomsToCorridors = true,
    calculateDirections = true
  } = options;
  
  const nodes = [];
  const edges = [];
  const roomMappings = [];
  const warnings = [];
  const errors = [];
  
  try {
    // Step 1: Place room nodes
    const roomNodes = placeRoomNodes(scanResult.rooms);
    nodes.push(...roomNodes);
    
    // Step 2: Place corridor nodes
    const corridorNodes = placeCorridorNodes(scanResult.paths);
    nodes.push(...corridorNodes);
    
    // Step 3: Merge nearby nodes (if enabled)
    if (mergeNearbyNodes) {
      mergeNodes(nodes, NODE_MERGE_THRESHOLD);
    }
    
    // Step 4: Detect and mark intersections
    markIntersections(nodes);
    
    // Step 5: Create corridor edges
    const corridorEdges = createCorridorEdges(scanResult.paths, nodes, calculateDirections);
    edges.push(...corridorEdges);
    
    // Step 6: Create room-to-corridor edges
    if (connectRoomsToCorridors) {
      const roomEdges = createRoomToCorridorEdges(roomNodes, corridorNodes, nodes, calculateDirections);
      edges.push(...roomEdges);
    }
    
    // Step 7: Create intersection edges
    const intersectionEdges = createIntersectionEdges(nodes, calculateDirections);
    edges.push(...intersectionEdges);
    
    // Step 8: Generate room mappings
    roomNodes.forEach(roomNode => {
      if (roomNode.metadata && roomNode.metadata.roomId) {
        roomMappings.push({
          roomId: roomNode.metadata.roomId,
          nodeId: roomNode.id,
          position: roomNode.position
        });
      }
    });
    
    // Step 9: Normalize and sort
    normalizeGraph(nodes, edges);
    sortGraph(nodes, edges);
    
    // Step 10: Build connections array for each node
    buildNodeConnections(nodes, edges);
    
    // Step 11: Calculate bounds
    const bounds = calculateGraphBounds(nodes);
    
    // Step 12: Validate graph
    const validation = validateGraph(nodes, edges, roomMappings);
    warnings.push(...validation.warnings);
    errors.push(...validation.errors);
    
    return {
      metadata: {
        source: scanResult.metadata?.source || 'unknown',
        timestamp: new Date().toISOString(),
        version: GRAPH_BUILDER_VERSION,
        nodeCount: nodes.length,
        edgeCount: edges.length,
        roomCount: roomMappings.length
      },
      nodes,
      edges,
      roomMappings,
      bounds,
      warnings,
      errors
    };
    
  } catch (error) {
    errors.push(`Graph building error: ${error.message}`);
    return {
      metadata: {
        source: scanResult.metadata?.source || 'unknown',
        timestamp: new Date().toISOString(),
        version: GRAPH_BUILDER_VERSION,
        nodeCount: 0,
        edgeCount: 0,
        roomCount: 0
      },
      nodes: [],
      edges: [],
      roomMappings: [],
      bounds: { x: 0, y: 0, width: 0, height: 0 },
      warnings,
      errors
    };
  }
}

/**
 * Place nodes for each room
 */
function placeRoomNodes(rooms) {
  const nodes = [];
  
  rooms.forEach((room, index) => {
    // Calculate room center
    const center = {
      x: room.bounds.x + room.bounds.width / 2,
      y: room.bounds.y + room.bounds.height / 2
    };
    
    const nodeType = determineNodeTypeFromRoom(room.type);
    const nodeId = generateNodeID(center, nodeType, index);
    
    nodes.push({
      id: nodeId,
      type: nodeType,
      position: normalizePoint(center),
      connections: [], // Will be populated later
      metadata: {
        roomId: room.id,
        label: room.label
      }
    });
  });
  
  return nodes;
}

/**
 * Place nodes along corridors
 */
function placeCorridorNodes(paths) {
  const nodes = [];
  const nodeMap = new Map(); // To avoid duplicates
  
  paths.forEach((path, pathIndex) => {
    path.segments.forEach((segment, segIndex) => {
      // Create nodes at segment endpoints
      const startNode = createCorridorNode(segment.start, path.id, `start_${segIndex}`, nodeMap);
      const endNode = createCorridorNode(segment.end, path.id, `end_${segIndex}`, nodeMap);
      
      if (startNode) nodes.push(startNode);
      if (endNode) nodes.push(endNode);
    });
  });
  
  return Array.from(nodeMap.values());
}

/**
 * Create a corridor node (with deduplication)
 */
function createCorridorNode(position, pathId, suffix, nodeMap) {
  const normalized = normalizePoint(position);
  const nodeId = generateNodeID(normalized, 'corridor', `${pathId}_${suffix}`);
  
  // Check if node already exists at this position
  const existingNode = findNodeAtPosition(nodeMap, normalized, NODE_MERGE_THRESHOLD);
  if (existingNode) {
    return null; // Node already exists
  }
  
  const node = {
    id: nodeId,
    type: 'corridor',
    position: normalized,
    connections: [],
    metadata: {
      pathId: pathId
    }
  };
  
  nodeMap.set(nodeId, node);
  return node;
}

/**
 * Find node at or near position
 */
function findNodeAtPosition(nodeMap, position, threshold) {
  for (const node of nodeMap.values()) {
    const distance = euclideanDistance(node.position, position);
    if (distance <= threshold) {
      return node;
    }
  }
  return null;
}

/**
 * Merge nearby nodes
 */
function mergeNodes(nodes, threshold) {
  const merged = new Set();
  const toRemove = [];
  
  for (let i = 0; i < nodes.length; i++) {
    if (merged.has(i)) continue;
    
    const node1 = nodes[i];
    const cluster = [node1];
    
    // Find all nodes within threshold
    for (let j = i + 1; j < nodes.length; j++) {
      if (merged.has(j)) continue;
      
      const node2 = nodes[j];
      const distance = euclideanDistance(node1.position, node2.position);
      
      if (distance <= threshold) {
        cluster.push(node2);
        merged.add(j);
        toRemove.push(j);
      }
    }
    
    // If cluster found, merge to first node
    if (cluster.length > 1) {
      // Update first node position to average
      const avgX = cluster.reduce((sum, n) => sum + n.position.x, 0) / cluster.length;
      const avgY = cluster.reduce((sum, n) => sum + n.position.y, 0) / cluster.length;
      node1.position = normalizePoint({ x: avgX, y: avgY });
      
      // Merge metadata
      const allPathIds = new Set();
      cluster.forEach(n => {
        if (n.metadata?.pathId) allPathIds.add(n.metadata.pathId);
        if (n.metadata?.roomId) {
          node1.metadata = node1.metadata || {};
          node1.metadata.roomId = n.metadata.roomId;
        }
      });
      if (allPathIds.size > 0) {
        node1.metadata = node1.metadata || {};
        node1.metadata.pathIds = Array.from(allPathIds);
      }
      
      // Mark as intersection if multiple paths
      if (allPathIds.size > 1) {
        node1.type = 'intersection';
        node1.metadata.isIntersection = true;
      }
    }
  }
  
  // Remove merged nodes (in reverse order to maintain indices)
  toRemove.sort((a, b) => b - a);
  toRemove.forEach(index => nodes.splice(index, 1));
}

/**
 * Mark intersection nodes
 */
function markIntersections(nodes) {
  // Nodes with multiple pathIds are intersections
  nodes.forEach(node => {
    if (node.metadata?.pathIds && node.metadata.pathIds.length > 1) {
      node.type = 'intersection';
      node.metadata.isIntersection = true;
    }
  });
}

/**
 * Create edges along corridors
 */
function createCorridorEdges(paths, allNodes, calculateDirections) {
  const edges = [];
  const edgeMap = new Map(); // For deduplication
  
  paths.forEach(path => {
    path.segments.forEach(segment => {
      const startPos = normalizePoint(segment.start);
      const endPos = normalizePoint(segment.end);
      
      // Find nodes at segment endpoints
      const startNode = findNearestNode(allNodes, startPos, NODE_MERGE_THRESHOLD * 2);
      const endNode = findNearestNode(allNodes, endPos, NODE_MERGE_THRESHOLD * 2);
      
      if (startNode && endNode && startNode.id !== endNode.id) {
        const distance = euclideanDistance(startNode.position, endNode.position);
        
        if (distance >= MIN_EDGE_DISTANCE) {
          const edgeId = generateEdgeID(startNode.id, endNode.id);
          
          // Check if edge already exists
          if (!edgeMap.has(edgeId)) {
            const edge = {
              id: edgeId,
              from: startNode.id,
              to: endNode.id,
              distance: roundToPrecision(distance),
              bidirectional: true,
              metadata: {
                pathId: path.id,
                type: path.type
              }
            };
            
            if (calculateDirections) {
              edge.direction = calculateDirection(startNode.position, endNode.position);
            }
            
            edges.push(edge);
            edgeMap.set(edgeId, edge);
          }
        }
      }
    });
  });
  
  return edges;
}

/**
 * Create edges connecting rooms to corridors
 */
function createRoomToCorridorEdges(roomNodes, corridorNodes, allNodes, calculateDirections) {
  const edges = [];
  const edgeMap = new Map();
  
  roomNodes.forEach(roomNode => {
    // Find nearest corridor node
    let nearestCorridorNode = null;
    let minDistance = Infinity;
    
    corridorNodes.forEach(corridorNode => {
      const distance = euclideanDistance(roomNode.position, corridorNode.position);
      if (distance < minDistance && distance <= ROOM_CORRIDOR_DISTANCE) {
        minDistance = distance;
        nearestCorridorNode = corridorNode;
      }
    });
    
    if (nearestCorridorNode) {
      const distance = euclideanDistance(roomNode.position, nearestCorridorNode.position);
      
      if (distance >= MIN_EDGE_DISTANCE) {
        const edgeId = generateEdgeID(roomNode.id, nearestCorridorNode.id);
        
        if (!edgeMap.has(edgeId)) {
          const edge = {
            id: edgeId,
            from: roomNode.id,
            to: nearestCorridorNode.id,
            distance: roundToPrecision(distance),
            bidirectional: true,
            metadata: {
              type: 'doorway'
            }
          };
          
          if (calculateDirections) {
            edge.direction = calculateDirection(roomNode.position, nearestCorridorNode.position);
          }
          
          edges.push(edge);
          edgeMap.set(edgeId, edge);
        }
      }
    }
  });
  
  return edges;
}

/**
 * Create edges at intersections
 */
function createIntersectionEdges(nodes, calculateDirections) {
  const edges = [];
  const edgeMap = new Map();
  
  // Find intersection nodes
  const intersectionNodes = nodes.filter(n => n.type === 'intersection' || n.metadata?.isIntersection);
  
  // For each intersection, connect to nearby nodes
  intersectionNodes.forEach(intersectionNode => {
    nodes.forEach(node => {
      if (node.id === intersectionNode.id) return;
      
      const distance = euclideanDistance(intersectionNode.position, node.position);
      
      // Connect if within reasonable distance (corridor width)
      if (distance <= ROOM_CORRIDOR_DISTANCE && distance >= MIN_EDGE_DISTANCE) {
        const edgeId = generateEdgeID(intersectionNode.id, node.id);
        
        if (!edgeMap.has(edgeId)) {
          const edge = {
            id: edgeId,
            from: intersectionNode.id,
            to: node.id,
            distance: roundToPrecision(distance),
            bidirectional: true,
            metadata: {
              type: 'intersection'
            }
          };
          
          if (calculateDirections) {
            edge.direction = calculateDirection(intersectionNode.position, node.position);
          }
          
          edges.push(edge);
          edgeMap.set(edgeId, edge);
        }
      }
    });
  });
  
  return edges;
}

/**
 * Build connections array for each node
 */
function buildNodeConnections(nodes, edges) {
  // Reset connections
  nodes.forEach(node => {
    node.connections = [];
  });
  
  // Build connections from edges
  edges.forEach(edge => {
    const fromNode = nodes.find(n => n.id === edge.from);
    const toNode = nodes.find(n => n.id === edge.to);
    
    if (fromNode && !fromNode.connections.includes(edge.to)) {
      fromNode.connections.push(edge.to);
    }
    
    if (edge.bidirectional && toNode && !toNode.connections.includes(edge.from)) {
      toNode.connections.push(edge.from);
    }
  });
}

/**
 * Calculate cardinal direction from one point to another
 */
function calculateDirection(from, to) {
  const dx = to.x - from.x;
  const dy = to.y - from.y;
  const angle = Math.atan2(dy, dx) * 180 / Math.PI;
  
  // Normalize angle to 0-360
  const normalizedAngle = (angle + 360) % 360;
  
  // Convert to cardinal direction
  if (normalizedAngle >= 315 || normalizedAngle < 45) return 'east';
  if (normalizedAngle >= 45 && normalizedAngle < 135) return 'south';
  if (normalizedAngle >= 135 && normalizedAngle < 225) return 'west';
  return 'north';
}

/**
 * Determine node type from room type
 */
function determineNodeTypeFromRoom(roomType) {
  switch (roomType) {
    case 'stair': return 'stair';
    case 'elevator': return 'elevator';
    case 'exit': return 'exit';
    default: return 'room';
  }
}

/**
 * Generate deterministic node ID
 */
function generateNodeID(position, type, suffix) {
  const normalized = normalizePoint(position);
  return `NODE_${type}_${Math.round(normalized.x)}_${Math.round(normalized.y)}_${suffix}`.replace(/[^a-zA-Z0-9_]/g, '_');
}

/**
 * Generate deterministic edge ID
 */
function generateEdgeID(fromNodeId, toNodeId) {
  // Sort IDs for deterministic edge direction
  const [id1, id2] = [fromNodeId, toNodeId].sort();
  return `EDGE_${id1}_${id2}`.replace(/[^a-zA-Z0-9_]/g, '_');
}

/**
 * Find nearest node to a position
 */
function findNearestNode(nodes, position, maxDistance = Infinity) {
  let nearest = null;
  let minDistance = Infinity;
  
  nodes.forEach(node => {
    const distance = euclideanDistance(node.position, position);
    if (distance < minDistance && distance <= maxDistance) {
      minDistance = distance;
      nearest = node;
    }
  });
  
  return nearest;
}

/**
 * Calculate Euclidean distance between two points
 */
function euclideanDistance(p1, p2) {
  const dx = p2.x - p1.x;
  const dy = p2.y - p1.y;
  return Math.sqrt(dx * dx + dy * dy);
}

/**
 * Normalize point coordinates
 */
function normalizePoint(point) {
  return {
    x: roundToPrecision(point.x),
    y: roundToPrecision(point.y)
  };
}

/**
 * Round to fixed precision
 */
function roundToPrecision(value) {
  const factor = Math.pow(10, COORDINATE_PRECISION);
  return Math.round(value * factor) / factor;
}

/**
 * Normalize graph (coordinates)
 */
function normalizeGraph(nodes, edges) {
  nodes.forEach(node => {
    node.position = normalizePoint(node.position);
  });
  
  edges.forEach(edge => {
    edge.distance = roundToPrecision(edge.distance);
  });
}

/**
 * Sort graph for deterministic output
 */
function sortGraph(nodes, edges) {
  // Sort nodes by position (y, then x)
  nodes.sort((a, b) => {
    const yDiff = a.position.y - b.position.y;
    if (Math.abs(yDiff) < 1) {
      return a.position.x - b.position.x;
    }
    return yDiff;
  });
  
  // Sort edges by from node, then to node
  edges.sort((a, b) => {
    if (a.from !== b.from) {
      return a.from.localeCompare(b.from);
    }
    return a.to.localeCompare(b.to);
  });
}

/**
 * Calculate graph bounds
 */
function calculateGraphBounds(nodes) {
  if (nodes.length === 0) {
    return { x: 0, y: 0, width: 0, height: 0 };
  }
  
  let minX = Infinity, minY = Infinity;
  let maxX = -Infinity, maxY = -Infinity;
  
  nodes.forEach(node => {
    minX = Math.min(minX, node.position.x);
    minY = Math.min(minY, node.position.y);
    maxX = Math.max(maxX, node.position.x);
    maxY = Math.max(maxY, node.position.y);
  });
  
  return {
    x: roundToPrecision(minX),
    y: roundToPrecision(minY),
    width: roundToPrecision(maxX - minX),
    height: roundToPrecision(maxY - minY)
  };
}

/**
 * Validate graph
 */
function validateGraph(nodes, edges, roomMappings) {
  const warnings = [];
  const errors = [];
  
  // Check for duplicate node IDs
  const nodeIds = new Set();
  nodes.forEach(node => {
    if (nodeIds.has(node.id)) {
      errors.push(`Duplicate node ID: ${node.id}`);
    }
    nodeIds.add(node.id);
  });
  
  // Check for duplicate edge IDs
  const edgeIds = new Set();
  edges.forEach(edge => {
    if (edgeIds.has(edge.id)) {
      errors.push(`Duplicate edge ID: ${edge.id}`);
    }
    edgeIds.add(edge.id);
  });
  
  // Check edges reference valid nodes
  const validNodeIds = new Set(nodes.map(n => n.id));
  edges.forEach(edge => {
    if (!validNodeIds.has(edge.from)) {
      errors.push(`Edge references invalid from node: ${edge.from}`);
    }
    if (!validNodeIds.has(edge.to)) {
      errors.push(`Edge references invalid to node: ${edge.to}`);
    }
  });
  
  // Check room mappings reference valid nodes
  roomMappings.forEach(mapping => {
    if (!validNodeIds.has(mapping.nodeId)) {
      errors.push(`Room mapping references invalid node: ${mapping.nodeId}`);
    }
  });
  
  // Check for isolated nodes
  const isolatedNodes = nodes.filter(node => node.connections.length === 0);
  if (isolatedNodes.length > 0) {
    warnings.push(`${isolatedNodes.length} isolated node(s) found (no connections)`);
  }
  
  // Check for rooms without mappings
  const mappedRoomIds = new Set(roomMappings.map(m => m.roomId));
  // This would require access to original rooms, so we'll skip for now
  
  return { warnings, errors };
}

// Export for Node.js / CommonJS
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    buildNavigationGraph,
    GRAPH_BUILDER_VERSION
  };
}

// Export for ES6 modules
if (typeof window !== 'undefined') {
  window.NavigationGraphBuilder = {
    buildNavigationGraph,
    GRAPH_BUILDER_VERSION
  };
}

File: AI/graph_builder.py
"""
Navigation Graph Builder Module

Converts floor plan detection results into a navigation graph suitable
for pathfinding and indoor navigation.
"""

import numpy as np
from typing import List, Dict, Tuple, Optional
import uuid


def generate_id(prefix: str = 'node') -> str:
    """Generate a unique ID for graph elements."""
    return f"{prefix}_{uuid.uuid4().hex[:8]}"


def build_navigation_graph(
    walls: List[Dict],
    rooms: List[Dict],
    doors: List[Dict],
    options: Dict = None
) -> Dict:
    """
    Build a complete navigation graph from detected floor plan elements.
    
    Args:
        walls: Detected walls list
        rooms: Detected rooms list
        doors: Detected doors list
        options: Graph building options:
            - floor: Floor number (default: 1)
            - corridor_spacing: Node spacing in corridors (default: 50)
            - max_edge_distance: Max distance for auto connections (default: 100)
            
    Returns:
        Navigation graph dictionary with nodes and edges
    """
    if options is None:
        options = {}
    
    floor = options.get('floor', 1)
    corridor_spacing = options.get('corridor_spacing', 50)
    max_edge_distance = options.get('max_edge_distance', 100)
    
    nodes = []
    edges = []
    
    # Create room nodes
    for room in rooms:
        center_x = room['x'] + room['width'] / 2
        center_y = room['y'] + room['height'] / 2
        
        node = {
            'id': generate_id('room'),
            'type': 'room',
            'label': room.get('name', 'Room'),
            'floor': floor,
            'position': {
                'x': float(center_x),
                'y': float(center_y)
            },
            'bounds': {
                'x': float(room['x']),
                'y': float(room['y']),
                'width': float(room['width']),
                'height': float(room['height'])
            },
            'properties': {
                'area': float(room.get('area', 0)),
                'confidence': float(room.get('confidence', 0.75))
            }
        }
        nodes.append(node)
    
    # Create door nodes
    door_nodes = []
    for door in doors:
        node = {
            'id': generate_id('door'),
            'type': 'door',
            'label': 'Door',
            'floor': floor,
            'position': {
                'x': float(door['hinge']['x']),
                'y': float(door['hinge']['y'])
            },
            'properties': {
                'swing_angle': float(door.get('swing_angle', 90)),
                'radius': float(door.get('radius', 30)),
                'confidence': float(door.get('confidence', 0.7))
            }
        }
        nodes.append(node)
        door_nodes.append(node)
    
    # Create corridor/path nodes based on areas between rooms
    corridor_nodes = create_corridor_nodes(rooms, walls, floor, corridor_spacing)
    nodes.extend(corridor_nodes)
    
    # Create edges between nearby nodes
    all_nodes = nodes.copy()
    
    for i, node in enumerate(all_nodes):
        pos1 = (node['position']['x'], node['position']['y'])
        
        for other in all_nodes[i + 1:]:
            pos2 = (other['position']['x'], other['position']['y'])
            
            dist = distance(pos1, pos2)
            
            # Auto-connect nodes within threshold
            if dist < max_edge_distance:
                # Check if connection is valid (not through a wall)
                if not crosses_wall(pos1, pos2, walls):
                    edge = {
                        'id': generate_id('edge'),
                        'from': node['id'],
                        'to': other['id'],
                        'weight': float(dist),
                        'type': determine_edge_type(node, other)
                    }
                    edges.append(edge)
    
    # Connect doors to nearby rooms
    for door_node in door_nodes:
        door_pos = (door_node['position']['x'], door_node['position']['y'])
        
        # Find closest room nodes
        for room_node in [n for n in nodes if n['type'] == 'room']:
            room_pos = (room_node['position']['x'], room_node['position']['y'])
            dist = distance(door_pos, room_pos)
            
            if dist < max_edge_distance * 1.5:
                # Add edge if not already connected
                existing = any(
                    (e['from'] == door_node['id'] and e['to'] == room_node['id']) or
                    (e['from'] == room_node['id'] and e['to'] == door_node['id'])
                    for e in edges
                )
                
                if not existing:
                    edges.append({
                        'id': generate_id('edge'),
                        'from': door_node['id'],
                        'to': room_node['id'],
                        'weight': float(dist),
                        'type': 'door_connection'
                    })
    
    # Remove duplicate edges
    edges = remove_duplicate_edges(edges)
    
    return {
        'version': '2.0',
        'floor': floor,
        'nodes': nodes,
        'edges': edges,
        'metadata': {
            'total_rooms': len([n for n in nodes if n['type'] == 'room']),
            'total_doors': len([n for n in nodes if n['type'] == 'door']),
            'total_connections': len(edges)
        }
    }


def create_corridor_nodes(
    rooms: List[Dict],
    walls: List[Dict],
    floor: int,
    spacing: int
) -> List[Dict]:
    """
    Create navigation nodes in corridor areas (spaces between rooms).
    
    Args:
        rooms: Detected rooms
        walls: Detected walls
        floor: Floor number
        spacing: Node spacing in pixels
        
    Returns:
        List of corridor nodes
    """
    nodes = []
    
    # Simple approach: create nodes at wall intersections and endpoints
    # that are not inside rooms
    
    for wall in walls:
        # Check midpoint of each wall
        mid_x = (wall['x1'] + wall['x2']) / 2
        mid_y = (wall['y1'] + wall['y2']) / 2
        
        # Skip if inside a room
        in_room = False
        for room in rooms:
            if point_in_rect(mid_x, mid_y, room):
                in_room = True
                break
        
        if not in_room:
            # Create a corridor node near the wall
            # Offset perpendicular to wall
            dx = wall['x2'] - wall['x1']
            dy = wall['y2'] - wall['y1']
            length = np.sqrt(dx * dx + dy * dy)
            
            if length > 0:
                # Perpendicular offset
                px = -dy / length * 30
                py = dx / length * 30
                
                node = {
                    'id': generate_id('path'),
                    'type': 'path',
                    'label': 'Corridor',
                    'floor': floor,
                    'position': {
                        'x': float(mid_x + px),
                        'y': float(mid_y + py)
                    },
                    'properties': {
                        'auto_generated': True
                    }
                }
                nodes.append(node)
    
    return nodes


def distance(p1: Tuple[float, float], p2: Tuple[float, float]) -> float:
    """Calculate Euclidean distance between two points."""
    return np.sqrt((p2[0] - p1[0]) ** 2 + (p2[1] - p1[1]) ** 2)


def point_in_rect(x: float, y: float, rect: Dict) -> bool:
    """Check if point is inside a rectangle."""
    return (rect['x'] <= x <= rect['x'] + rect['width'] and
            rect['y'] <= y <= rect['y'] + rect['height'])


def crosses_wall(
    p1: Tuple[float, float],
    p2: Tuple[float, float],
    walls: List[Dict]
) -> bool:
    """
    Check if a line segment crosses any wall.
    
    Args:
        p1: Start point
        p2: End point
        walls: List of walls
        
    Returns:
        True if the segment crosses a wall
    """
    for wall in walls:
        wall_p1 = (wall['x1'], wall['y1'])
        wall_p2 = (wall['x2'], wall['y2'])
        
        if segments_intersect(p1, p2, wall_p1, wall_p2):
            return True
    
    return False


def segments_intersect(
    p1: Tuple[float, float],
    p2: Tuple[float, float],
    p3: Tuple[float, float],
    p4: Tuple[float, float]
) -> bool:
    """Check if two line segments intersect."""
    def ccw(A, B, C):
        return (C[1] - A[1]) * (B[0] - A[0]) > (B[1] - A[1]) * (C[0] - A[0])
    
    return (ccw(p1, p3, p4) != ccw(p2, p3, p4) and 
            ccw(p1, p2, p3) != ccw(p1, p2, p4))


def determine_edge_type(node1: Dict, node2: Dict) -> str:
    """Determine edge type based on connected node types."""
    types = {node1['type'], node2['type']}
    
    if 'door' in types:
        return 'door_connection'
    elif types == {'room'}:
        return 'room_to_room'
    elif 'path' in types:
        return 'corridor'
    else:
        return 'connection'


def remove_duplicate_edges(edges: List[Dict]) -> List[Dict]:
    """Remove duplicate edges (same from/to or to/from)."""
    seen = set()
    unique = []
    
    for edge in edges:
        key = tuple(sorted([edge['from'], edge['to']]))
        if key not in seen:
            seen.add(key)
            unique.append(edge)
    
    return unique
